keps/sig-scheduling/20190317-non-preempting-priorityclass.md

```diff
@@ -0,0 +1,116 @@
+---
+title: Add NonPreempting Option For PriorityClasses
+authors:
+  - "@vllry"
+owning-sig: sig-scheduling
+participating-sigs:
+  - sig-scheduling
+reviewers:
+  - "k82cn"
+  - "wgliang"
+approvers:
+  - "bsalamat"
+editor: Vallery Lancey
+creation-date: 2019-03-17
+last-updated: 2019-03-17
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Allow PriorityClasses To Be Non-Preempting
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+[PriorityClasses](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/) are a GA feature as on 1.14,
+which impact the scheduling and eviction of pods.
+Pods will be scheduled according to descending priority.
+If a pod cannot be scheduled due to insufficient resources,
+lower-priority pods will be descheduled ("preempted") to make room.
+
+This proposal makes the preempting behavior optional for a PriorityClass,
+by adding a new field to PriorityClasses.
+If a PriorityClass does not have preemption enabled,
+a pod of that PriorityClass will not trigger preemption of other pods.
+
+## Motivation
+
+High-priority, non-preempting workloads are a common data science use case.
+Preempting batch workloads is a waste, as the work unit must be repeated.
+
+### Goals
+
+Add a boolean to PriorityClasses,
+to enable or disable preemption for pods of that PriorityClass.
+
+### Non-Goals
+
+## Proposal
+
+Add a Preempting field to PriorityClasses.
+This field will default to true,
+for backwards compatibility.
+
+```
+type PriorityClass struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+	Value int32
+	GlobalDefault bool
+	Description string
+	// New option
+	Preempting bool
```

Comment:
a bool pointer here？

Reply:
This paste is for the PriorityClass (which doesn't use a pointer)

Reply:
But how can we set the default as true ?

Reply:
In defaults.go [(reference)](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#edit-defaultsgo) you can convert nil/unset to true.

Reply:
defaulting an optional field while in alpha doesn't work well (since we don't want to persist or return the field of the alpha feature is disabled). this will need to be a *bool so we can correctly omit it until promoted

Reply:
I think if we want to distinguish between an unset value and an automatic false value of bool, we also need *bool?

Reply:
Thanks for the context @liggitt, that makes sense. I haven't worked with the API before.